app/services/first_registration/prospective_service.rb

```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
+module FirstRegistration
+  class ProspectiveService
+    attr_reader :prospective
+
+    def initialize(prospective)
+      @prospective = prospective
+      extend "FirstRegistration::#{@prospective.class.name}Service".safe_constantize
```

Comment:
Önemine binaen bu kısmı biraz törenle yapalım mı? (isimlendirmeyi sana bırakıyorum)

```ruby
    def initialize(prospective)
      @prospective = prospective
      specialize
    end

    private

    def specialize
      extend "FirstRegistration::#{prospective.class.name}Service".safe_constantize
    end
```

Reply:
@ecmelkytz @huseyin'in de dikkat çekmesiyle bu koda biraz daha yakından baktım.  Nesnenin çalışma zamanında (runtime) extend edilmesi zaman zaman benim de kullandığım bir teknik.  Ama bunu yaparken bir kaç şeye dikkat etmeye çalışıyorum:

- Sadece ufak davranış farklılıkları için yeni sınıflar oluşturmamak adına yapıyorum bunu.  Davranış farklılığı dediğim bir veya bilemedin iki metodla temsil ediliyor.  Bunları ilgili sınıfta gizlediğim modüllere aldıktan sonra nesneyi extend yapıyorum.

- Extend'i doğrudan nesnenin kendisine yaptırmak yerine daha üstte nesne dışında yapmaya çalışıyorum.  Şunun gibi:

```ruby
nesne = Nesne.new
nesne.extend Farklılıklar
```

Burada yapılan extend bana çok "helal" gelmedi.  Yukarıdaki ilkeler + şu nedenlerle:

- Buradaki extend küçük davranış farklılığından ziyade "ontolojik" bir farklılığı kotarmaya çalışıyor.  Extend edilen modüllere bakın, bunlar hiç de küçük değil.

- Extend'e rağmen `ProspectiveEmployeeService` ve `ProspectiveStudentService` modülleri DRY değil.  Epey ortaklıklar var.

- Extend yerine daha düzgün bir NYP yaklaşımıyla çözülebilir bu sorun (örnekleyeceğim).  Böyle bir yol varsa daima runtime extend'e tercih ederim.

Bir sonraki yorumda bir taslak çıkaracağım.  @huseyin sen de çalışır mısın bu taslağın gerçeklemesi üzerinde?

Reply:
`ProspectiveEmployeeService` ve `ProspectiveStudentService` modüllerine baktığımda temel ayrımın `Employee.new` ve `Student.new` çağrıları olduğunu görüyorum.  Bu modüllerde kalan diğer metodlar her ikisinde ortak zaten; buna sonra geleceğim.

Bence bu `.new` çağrılarını sırasıyla `Employee` ve `Student` modellerine sınıf metodu olarak taşıyalım.  Şuna benzer (Employee için örnekledim, Student için uyarlayın):

```ruby
class Employee < ApplicationRecord
  ...    
  def self.build_from_prospective_user(prospective, user)
    new(
      user: user,
      title_id: prospective.title_id,
      staff_number: prospective.staff_number
    )
  end
```

Yani Prospective'ten somut (concrete) Employee oluşturma lojiğini/sorumluluğunu Employee modeline verdim.

Sonra buraya dönelim.  Sadece `ProspectiveService` sınıfı yeterli olacak gibi.  Şu farkla sınıfa yeni bir değişken, örneğin `concrete` ekleyelim.  Bu değişken `Employee` veya `Student` kayıtlarına karşı geliyor.

```ruby
module FirstRegistration
  class ProspectiveService
    attr_reader :prospective

    def initialize(prospective)
      @prospective = prospective
      @concrete    = concrete_build
    end

    ...

    private

    attr_reader :concrete

    def concrete_build
      # find klass from prospective
      # invoke klass.build_from_prospective_user
    end

    ...
end
```

Sonra da dönüp önceki modüllerde bulunan ortak metodları `concrete`'i dikkate alarak ekleyelim:

```ruby
module FirstRegistration
  class ProspectiveService
    def valid?
      [prospective, concrete, user].all?(&:valid?)
    end

    def register
      return set_error_messages unless valid?

      user.save
      concrete.save
    end

    def error_messages
      concrete.errors.messages.merge(user.errors.messages)
    end
  end
end
```

Bu kabaca şimdi aklıma gelen çözüm.  Daha farklı çözümler bulunabilir.

Reply:
Bu önerdiğim düzenlemeyle `prospective_*_service.rb`'ler (iki dosya) kalkıyor, `prospective_service.rb` çok az şişmanlıyor.  Ayrıca "concern separation" daha iyi bir seviyede oluyor.  Ama oturup yazmadım, kurguya uymayan şeyler çıkabilir.

Reply:
> Bence bu .new çağrılarını sırasıyla Employee ve Student modellerine sınıf metodu olarak taşıyalım. Şuna benzer (Employee için örnekledim, Student için uyarlayın)

`ProspectiveEmployee` ve `ProspectiveStudent` daha uygun görünüyor.  Şöyle:

```ruby
class ProspectiveEmployee < ApplicationRecord
  ...    
  def self.build_concrete_user
    new(
      user: user,
      title_id: title_id,
      staff_number: staff_number
    )
  end
```

Reply:
Sınıf metodu mu olur nesne metodu mu olur, hangi argümanları ister vs bunu size bırakayım.

Reply:
`concrete` yerine daha uygun bir isim bulabilirsiniz: `registered` vs gibi.  Bunda kritik olan prospektiflikten çıkmış resmi `Employee` ve `Student`'ı temsil eden ortak bir şey olması.